Source/Microsoft.Gateway/Microsoft.InnerEye.Listener.Common/Providers/BaseConfigProvider.cs

```diff
@@ -25,110 +25,201 @@ public class BaseConfigProvider<T>
+            }
+
+            _fileSystemWatcher = new FileSystemWatcher(folderName)
+            {
+                Filter = !string.IsNullOrWhiteSpace(settingsFile) ? settingsFile : "*.json",
+                NotifyFilter = NotifyFilters.LastWrite,
+            };
+
+            _fileSystemWatcher.Changed += OnChanged;
+            _fileSystemWatcher.EnableRaisingEvents = true;
+
+            Load();
         }
 
         /// <summary>
-        /// Load T or Ts from a JSON file or folder.
+        /// File watcher Changed event handler. Filter the events, reload the config and if successful invoke ConfigChanged.
         /// </summary>
-        protected void Load()
+        /// <param name="sender">Sender.</param>
+        /// <param name="e">File system event args.</param>
+        private void OnChanged(object sender, FileSystemEventArgs e)
+        {
+            if (e.ChangeType != WatcherChangeTypes.Changed)
+            {
+                return;
+            }
+
+            var logEntry = LogEntry.Create(ServiceStatus.NewConfigurationDetetected,
+                string.Format("Settings have changed: {0}", e.FullPath));
+            logEntry.Log(_logger, LogLevel.Information);
+
+            if (!Load())
+            {
+                return;
+            }
+
+            ConfigChanged?.Invoke(this, new EventArgs());
+        }
+
+        /// <summary>
+        /// Load T from a JSON file or folder.
+        /// </summary>
+        /// <returns>True if new config has been loaded, false otherwise.</returns>
+        private bool Load()
         {
             if (File.Exists(_settingsFileOrFolderName))
             {
-                _ts = null;
+                var (t, loaded, parsed) = LoadFile(_settingsFileOrFolderName);
+
+                if (!loaded || !parsed)
+                {
+                    return false;
+                }
+
+                Config = t;
 
-                (_t, _) = LoadFile(_settingsFileOrFolderName);
+                return true;
             }
             else if (Directory.Exists(_settingsFileOrFolderName))
             {
-                _t = default(T);
                 var ts = new List<T>();
 
                 foreach (var file in Directory.EnumerateFiles(_settingsFileOrFolderName, "*.json"))
                 {
-                    var (t, loaded) = LoadFile(file);
-                    if (loaded)
+                    var (t, loaded, parsed) = LoadFile(file);
+                    if (!loaded)
+                    {
+                        // File still in use, FileWatcher has reported file changed but
+                        // the other process has not finished yet.
+                        return false;
+                    }
+
+                    if (parsed)
                     {
                         ts.Add(t);
                     }
                 }
 
-                _ts = ts.ToArray();
+                Config = _flatMap.Invoke(ts);
```

Comment:
Do you need this Invoke or can you just call it with ()

Reply:
Invoke is not needed, I just prefer it for clarity.